pkg/finalizer/finalizer.go

```diff
@@ -0,0 +1,53 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package finalizer
+
+import (
+	"context"
+	"fmt"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// NewFinalizers returns the Finalizers interface
+func NewFinalizers() Finalizers {
+	return finalizers(map[string]Finalizer{})
+}
+
+func (f finalizers) Register(key string, finalizer Finalizer) error {
+	if _, ok := f[key]; ok {
+		return fmt.Errorf("finalizer for key %q already registered", key)
+	}
+	f[key] = finalizer
+	return nil
+}
+
+func (f finalizers) Finalize(ctx context.Context, obj client.Object) (bool, error) {
+	needsUpdate := false
+	for key, finalizer := range f {
+		if obj.GetDeletionTimestamp().IsZero() && !controllerutil.ContainsFinalizer(obj, key) {
```

Comment:
```suggestion
		if dt := obj.GetDeletionTimestamp(); (dt == nil || dt.IsZero()) && !controllerutil.ContainsFinalizer(obj, key) {
```

Reply:
I'm pretty sure IsZero has an `== nil` check in it already.

Reply:
If the `*time.Time` returned by `GetDeletionTimestamp()` is nil, then calling a method on it will panic.

Reply:
I don't think that's correct, https://play.golang.org/p/JtZblcMxSd9 seems to work for me. Pretty sure Go allows defining methods on null pointers like any other typed pointer, the value passed in will just be nil.

Reply:
If a method has to dereference a pointer receiver to run and the receiver is nil, then the method will panic. See https://play.golang.org/p/1VZxyB9uq-c. Since `IsZero()` is not implemented on a pointer, it will panic.

Reply:
GetDeletionTimestamp() returns a *metav1.Time. Our version of IsZero is defined with a pointer receiver https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/time.go#L59

Reply:
Ah silly me, read it as std time. Resolving this comment.